File: packages/senri_ffi_js/src/types/adapter.ts
/*
 * Copyright (c) 2026 TT23XR Studio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { NormalizedType } from './normalized';

export interface FFIAdapter {
  loadLibrary(path: string): any;
  bindFunction(libHandle: any, name: string, retType: NormalizedType, argTypes: NormalizedType[], options?: any): (...args: any[]) => any;
  closeLibrary(libHandle: any): void;

  alloc(size: number): any;
  free(ptr: any): void;
  addressOf(buffer: ArrayBuffer | ArrayBufferView): bigint;

  registerCallback(func: Function, retType: NormalizedType, argTypes: NormalizedType[]): any;
  unregisterCallback(ptr: any): void;

  getErrno(): number;
  getStrerror(errno: number): string;

  mapType(type: NormalizedType): any;
  createPointerType(ofType: NormalizedType): any;
  createArrayType(ofType: NormalizedType, length: number): any;
  createStructType(fields: Record<string, NormalizedType>, packed?: number, size?: number, align?: number): any;
}

File: packages/senri_ffi_py/src/senri_ffi/types/adapter.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Any

from senri_ffi.types.normalized import NormalizedType


class FFIAdapter(ABC):
    @abstractmethod
    def load_library(self, path: str) -> Any: ...

    @abstractmethod
    def bind_function(
        self,
        lib_handle: Any,
        name: str,
        ret_type: NormalizedType,
        arg_types: list[NormalizedType],
        options: Any = None,
    ) -> Any: ...

    @abstractmethod
    def close_library(self, lib_handle: Any) -> None: ...

    @abstractmethod
    def alloc(self, size: int) -> Any: ...

    @abstractmethod
    def free(self, ptr: Any) -> None: ...

    @abstractmethod
    def address_of(self, buffer: Any) -> int: ...

    @abstractmethod
    def register_callback(
        self,
        func: Any,
        ret_type: NormalizedType,
        arg_types: list[NormalizedType],
    ) -> Any: ...

    @abstractmethod
    def unregister_callback(self, ptr: Any) -> None: ...

    @abstractmethod
    def get_errno(self) -> int: ...

    @abstractmethod
    def get_strerror(self, errno: int) -> str: ...

    @abstractmethod
    def map_type(self, t: NormalizedType) -> Any: ...

    @abstractmethod
    def create_pointer_type(self, of_type: NormalizedType) -> Any: ...

    @abstractmethod
    def create_array_type(self, of_type: NormalizedType, length: int) -> Any: ...

    @abstractmethod
    def create_struct_type(
        self,
        fields: dict[str, NormalizedType],
        packed: int | None = None,
        size: int = 0,
        align: int = 1,
    ) -> Any: ...
